Add mobile OfflineStatus component

derekblank opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
Adds a mobile component to display a visual indicator when working offline within the mobile editor. (Note: the functionality to subscribe to network status will be handled in a separate PR.) 

## Why?
Overall, to improve the user's experience when uploading media. 

Addresses:
* https://github.com/wordpress-mobile/gutenberg-mobile/issues/6405

## How?
Adds OfflineStatus component and associated icon/style files. 

## Testing Instructions
Applying the test patch will display the component within the editor when offline.

<details><summary>Test Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/block-list/index.native.js b/packages/block-editor/src/components/block-list/index.native.js
index 810e23e4c1..19540eddb4 100644
--- a/packages/block-editor/src/components/block-list/index.native.js
+++ b/packages/block-editor/src/components/block-list/index.native.js
@@ -28,6 +28,7 @@ import {
 	DEFAULT_BLOCK_LIST_CONTEXT,
 } from './block-list-context';
 import { BlockDraggableWrapper } from '../block-draggable';
+import OfflineStatus from '../offline-status';
 import { useEditorWrapperStyles } from '../../hooks/use-editor-wrapper-styles';
 import { store as blockEditorStore } from '../../store';
 
@@ -235,6 +236,7 @@ export default function BlockList( {
 			onLayout={ onLayout }
 			testID="block-list-wrapper"
 		>
+			<OfflineStatus />
 			{ isRootList ? (
 				<BlockListProvider
 					value={ {
```

</details>


## Screenshots or screencast <!-- if applicable -->
Light | Dark
-|-
<img width="333" alt="Screenshot 2023-12-07 at 6 18 07 pm" src="https://github.com/WordPress/gutenberg/assets/643285/d0633040-0057-4f8d-91e1-db5c98ca4c16"> | <img width="338" alt="Screenshot 2023-12-07 at 6 17 02 pm" src="https://github.com/WordPress/gutenberg/assets/643285/ccc1ee42-c883-4a98-bb26-aa3aa0398c14">

derekblank committed:
Add mobile OfflineStatus component

ellatrix committed:
Revert format types hook refactor (#56859)

Commit:
Adding `aria-sort` to table view headers (#56860)

Ensures sorting semantics are not just communicated visually

dcalhoun commented on packages/block-editor/src/components/offline-status/index.native.js:
```diff
@@ -0,0 +1,46 @@
+		styles[ 'icon--dark' ]
+	);
+
+	return ! isConnected ? (
```
If you believe it to be most effective, we could go ahead and implement this testing diff shared in this PR. Adding a `__DEV__` check to this conditional would mean we would only need to remove the the `__DEV__` flag to ship this feature. 

Not necessarily requesting this be done, merely sharing it for thought.

derekblank replied:
Updated to add the component behind a `__DEV__` flag, and rectified the commit history. 👍

dcalhoun commented on packages/react-native-bridge/android/react-native-bridge/src/main/java/org/wordpress/mobile/ReactNativeGutenbergBridge/GutenbergBridgeJS2Parent.java:
Rebasing this branch atop the latest `feat/connectivity-status` should get rid of this file diff and the others related to the `isConnected` Hook. A new PR is opened in https://github.com/WordPress/gutenberg/pull/56861.

dcalhoun reviewed (commented):
Overall, the changes look good and the status indicator correctly shows up when the network status changes. 🎉 

I noted the status appears invisible on Android when dark mode is enabled. Do you see this as well? 

<details><summary>Android Dark Mode Screenshot</summary>

![android dark mode status invisible](https://github.com/WordPress/gutenberg/assets/438664/8fb42afa-ec0c-4a94-982b-ab42f6a976e0)


</details> 

Also, I'm not sure if my eyes are playing tricks on me or if the status text and icon are _not_ vertically centered. WDYT? 

<details><summary>Status Vertically Centered</summary>

![offline status not vertically centered](https://github.com/WordPress/gutenberg/assets/438664/ae039b08-8f5f-4eb9-a8c4-4da9445a1899)


</details>

jameskoster committed:
Update data view layout (#56786)

Co-authored-by: ntsekouras <ntsekouras@outlook.com>

youknowriad committed:
Editor: Move the device type state to the editor package (#56866)

gutenbergplugin committed:
Update changelog files

gutenbergplugin committed:
chore(release): publish

 - @wordpress/block-directory@4.24.1
 - @wordpress/block-library@8.24.1
 - @wordpress/create-block-interactive-template@1.10.1
 - @wordpress/customize-widgets@4.24.1
 - @wordpress/e2e-tests@7.18.1
 - @wordpress/edit-post@7.24.1
 - @wordpress/edit-site@5.24.1
 - @wordpress/edit-widgets@5.24.1
 - @wordpress/editor@13.24.1
 - @wordpress/interactivity@3.0.1

chad1008 committed:
Tabs: Implement `ariakit/test` in unit tests (#56835)

youknowriad committed:
Site Editor: Fix active edited post (#56863)

colorful-tones committed:
Update InnerBlocks defaultblock doc usage (#56728)

Co-authored-by: Riad Benguella <benguella@gmail.com>

ellatrix committed:
One hook to rule them all: preparation for a block supports API (#56862)

ellatrix committed:
RichText: pass value to store (#43204)

brookewp committed:
`FontSizePicker`: Add opt-in prop for 40px default size (#56804)

* `FontSizePicker`: Add opt-in prop for 40px default size

* Update button when __next40pxDefaultSize is true

* Update changelog

ellatrix committed:
Keycodes: avoid regex for capital case (#56822)

jeryj committed:
Refactor <BlockToolbar />: Add variant prop and <NavigableToolbar /> wrapper (#56335)

- Combine `<BlockContextualToolbar />` into `<BlockToolbar />`. This brings all the necessary functionality for `<BlockToolbar />` to be used, mainly by wrapping it in a `<NavigableToolbar />`.
- Remove `<BlockContextualToolbar />`. This is no longer needed, and was originally intended for Contextual to mean Popover.
- Replace usage of `<BlockContextualToobar />` in edit-site, edit-post, edit-widget, etc headers with `<BlockToolbar />`
- Refactor `<SelectedBlockTools />` to become `<BlockToolbarPopover />` to better reflect its purpose.
- Add `<PrivateBlockToolbar />` with new props of `__experimentalInitialIndex`, `__experimentalOnIndexChange`, and `focusOnMount`.
- Export `<BlockToolbar />` from  `<PrivateBlockToolbar />` with locked `__experimentalInitialIndex`, `__experimentalOnIndexChange`, and `focusOnMount` as `undefined`. Public API props are `hideDragHandles` and `variant` with a default of `unstyled`.
- Remove concept of `isFixed` from the `<BlockToolbar />`. The styles of the current small-screen fixed toolbar will become the default. The popover and top toolbars will have its own CSS overrides applied. 
- Split `<BlockToolbarPopover />` and `<BlockBreadcrumbPopover />` into separate components for simplicity.
- Top Toolbar now means. You have to implement the `<BlockToolbar />` on your own.

ellatrix committed:
Mobile: remove rich text multiline (#56117)

jeryj committed:
Remove BlockTools BackCompat (#56874)

It was deprecated in 5.8 and slated to be removed in 6.3.

derekblank committed:
Add OfflineStatus component to block-list behind __DEV__ flag

brookewp committed:
`BorderControl`: Fix button styles (#56730)

* `BorderControl`: Fix button styles

* Add __next40pxDefaultSize to Button when BorderControl has large size

* Update changelog

* Remove explicit return leftover from unused changes

derekblank commented:
> Also, I'm not sure if my eyes are playing tricks on me or if the status text and icon are _not_ vertically centered. WDYT?

I think it's the icon that is not correct. I used a placeholder icon I found for free, but was intending to replace it after we received some design feedback. I'll seek that out now and replace it here as we're leaving the PR open.

ntsekouras committed:
DataViews: Remove TanStack (#56873)

getdave committed:
Simplify page list edit warning (#56829)

* Alter wording

* Make button name match intention

* Update wording

* Use tweaked wording

* Copy tweak + remove emphasis

* Tweak

---------

Co-authored-by: Rich Tabor <hi@richtabor.com>

dcalhoun commented on packages/block-editor/src/components/offline-status/index.native.js:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Internal dependencies
+ */
+import styles from './style';
```
Looks like my recent suggestion to drop the `native` suffix here resulted in test failures. It appears Metro correctly resolve the file when only a single, suffixed file is present. However, Jest fails resolution in that context. I recall encountering this before. We need to reinstate the `native` suffix here. 

Sorry to misguide you and for the inconvenience. ☹️

dcalhoun reviewed (commented):
> > I noted the status appears invisible on Android when dark mode is enabled. Do you see this as well?
> 
> I did not note this, but it appears like it may be theme-based as I noted the editor background is not dark in your screenshot. I can take a look, however. What theme were you using? 

Interesting. My screenshot is from a site using [Geologist](https://wordpress.org/themes/geologist/). 

My screenshot showcases not only an invisible background, but invisible text. Even if the theme is at play and the editor background is incorrect (concerning unto itself), shouldn't the text still be visible? The offline status UI has its own background and text colors, correct?

chad1008 committed:
Components: replace `TabPanel` with `Tabs` in the editor's `ColorPanel` (#56878)

* replace `TabPanel` with `Tabs`

* implement initial tab to track current value

* focusable false

* defer unlock call

miminari committed:
copy/fix capitalization of WordPress (#56834)

taylorgorman committed:
Link to Dashicons (#56872)

ellatrix committed:
Fix PHP linter failing (#56905)

ciampo committed:
DropdownMenuV2Ariakit: prevent prefix collapsing if all radios or checkboxes are unselected (#56720)

* DropdownMenuV2Ariakit: prevent prefix collapsing if all radios or checkboxes are unselected

* CHANGELOG

* DRY it up

chad1008 committed:
Tabs: replace `id` with new `tabId` prop (#56883)

* replace `id` with `tabId`

* update stories and tests

* update `ColorGradientControl` implementation

* changelog

* use `tabId` in tests for consistency

* update `ColorPanel` implementation

dcalhoun committed:
fix: Guard against a block styles crash due to null block values (#56903)

* fix: Guard against a block styles crash due to null block values

In certain scenarios, e.g., when the editor hangs due to poor
performance, the `getBlock` value unexpectedly returns `null`. To guard
against a crash in this scenario, we now conditionally access attributes
on the block.

A ideal resolution would be improving editor performance to avoid the
scenario where a race condition causes a crash, but the fact remains
that `getBlock` _can_ return `null` values and that we should not
presume that it will not.

* docs: Add change log entry

ellatrix committed:
Block editor: hooks: manage BlockListBlock filters in one place (#56875)

chad1008 committed:
Tabs: sync browser focus to selected tab in controlled mode (#56658)

* link browser focus to selected tab changes in controlled mode

* remove unnecessary memoization

* remove duplicate condition

* dont apply when selectOnMove is false

* use `items` to find active element, instead of a ref

* CHANGELOG

* add `selectOnMove` unit tests

* convert new tests to ariakit/test

* remove redundant tests

youknowriad committed:
Editor: Move the edit template blocks notification to editor package (#56901)

bph committed:
Doc: Spinner - add Storybook link (#56818)

* Add link to the Spinner's Storybook page

* remove 'also'

scruffian committed:
Improve text and design of the block removal warnings (#56869)

* Improve text and design of the block removal warnings

* Updating heading from "Are you sure?" to "Be careful!"

---------

Co-authored-by: annezazu <annezazu@gmail.com>

youknowriad committed:
Editor: Unify device preview styles (#56904)

apeatling committed:
Revert "Doc: Spinner - add Storybook link (#56818)" (#56913)

This reverts commit 25cd2468f825dc5bf1f92185aef27c5173d0d22e.

t-hamano committed:
Don't render undefined classname in useBlockProps hook (#56923)

* Remove undefined classname in useBlockProps hook

* Consider when block default classname is undefined

dmsnell committed:
HTML API: Backport updates from Core, fix Interactivity API

Fix: Update Interactivity API Use of Private HTML API Classes to avoid Crash

Previously the HTML API was using a mixture of approaches in its helper classes
for describing spans of text. One was (start, end) and the other was (start, length).
In WordPress/wordpress-develop#5721 these were all normalized to (start, length) but
the Interactivity API was not updated in concert with that change.

In this patch the `inner_html` methods are updated to use the appropriate formats.

this PR also includes the necessary back port of the HTML API to provide the new methods so that older WP installs don’t crash

derekblank committed:
Merge branch 'feat/connectivity-status' into rnmobile/connect-offline-status-component

derekblank committed:
Merge branch 'rnmobile/connect-offline-status-component' of https://github.com/WordPress/gutenberg into rnmobile/connect-offline-status-component

derekblank committed:
Replace offline icon and update OfflineStatus text alignment
